src/cfi.rs

```diff
@@ -454,12 +456,734 @@ impl<'input, Endian> FrameDescriptionEntry<'input, Endian>
+///
+/// ```
+/// use gimli::{UninitializedUnwindContext, UnwindTable};
+///
+/// # fn foo<'a>(some_fde: gimli::FrameDescriptionEntry<'a, gimli::LittleEndian>) -> gimli::Result<()> {
+/// // An uninitialized context.
+/// let ctx = UninitializedUnwindContext::new();
+///
+/// // Initialize the context by evaluating the CIE's initial instruction program.
+/// let mut ctx = try!(ctx.initialize(some_fde.cie()));
+///
+/// {
+///     // The initialized context can now be used to generate the unwind table.
+///     let mut table = UnwindTable::new(&mut ctx, &some_fde);
+///     while let Some(row) = try!(table.next_row()) {
+///         // Do stuff with each row...
+/// #       let _ = row;
+///     }
+/// }
+///
+/// // Reset the context to the uninitialized state and re-use it with other CFI
+/// // programs.
+/// let ctx = ctx.reset();
+/// # let _ = ctx;
+/// # Ok(())
+/// # }
+/// ```
+///
+/// In general, the states will flow from one to the other in accordance to the
+/// following diagram:
+///
+/// ```text
+///         +-------+
+///         | Start |
+///         +-------+
+///             |
+///             |
+/// UninitializedUnwindContext::new()
+///             |
+///             |
+///             V
+/// +----------------------------+
+/// | UninitializedUnwindContext |<---------------.
+/// +----------------------------+                |
+///             |                                 |
+///             |                                 |
+///    ctx.initialize(&cie)              Use with UnwindTable,
+///             |                        and then do ctx.reset()
+///             |                                 |
+///             V                                 |
+///  +--------------------------+                 |
+///  | InitializedUnwindContext |-----------------'
+///  +--------------------------+
+///             |
+///             |
+///            Drop
+///             |
+///             |
+///             V
+///          +-----+
+///          | End |
+///          +-----+
+/// ```
+#[derive(Clone, Debug)]
+pub struct UninitializedUnwindContext<'input, Endian>(UnwindContext<'input, Endian>)
+    where Endian: Endianity;
+
+impl<'input, Endian> UninitializedUnwindContext<'input, Endian>
+    where Endian: Endianity
+{
+    /// Construct a new call frame unwinding context.
+    pub fn new() -> UninitializedUnwindContext<'input, Endian> {
+        UninitializedUnwindContext(UnwindContext::new())
+    }
+
+    /// Run the CIE's initial instructions, creating an
+    /// `InitializedUnwindContext`.
+    pub fn initialize(mut self,
+                      cie: &CommonInformationEntry<'input, Endian>)
+                      -> Result<InitializedUnwindContext<'input, Endian>> {
+        assert!(self.0.is_uninitialized());
+
+        {
+            let mut table = UnwindTable::new_internal(&mut self.0, cie, None);
+            while let Some(_) = try!(table.next_row()) {}
+        }
```

Comment:
So we're evaluating the CIE initial_instructions for every FDE? It seems like avoiding that would be an important optimization if you want to evaluate instructions for multiple FDEs, and the current API doesn't let the user avoid it.